File: src/statesman/commands/SwitchConditionalCommand.java
package statesman.commands;

import statesman.Interpreter;

public class SwitchConditionalCommand extends ConditionalCommand {

    public static final String ID = "swcond";

    private int[] _switchIds;

    public SwitchConditionalCommand() {
        super();
    }
    
    public SwitchConditionalCommand(
            CommandGroup group,
            CommandGroup elseGroup,
            int[] switchIds,
            boolean[] targetValues,
            boolean orMode) {
        super();
        _group = group;
        _elseGroup = elseGroup;
        _switchIds = switchIds;
        _targetValues = targetValues;
        _orMode = orMode;
    }

    @Override
    public Command createInstance(String[] arguments) {
        if (arguments.length == 2) {
            String condition = arguments[1];
            boolean orMode = useOrOperator(condition);
            String[] parts = getConditionParts(condition, orMode);
            boolean[] targetValues = new boolean[parts.length];
            int[] switchIds = new int[parts.length];
            
            for (int i = 0; i < parts.length; i++) {
                targetValues[i] = !parts[i].startsWith("!");
                if (targetValues[i]) {
                    switchIds[i] = Integer.parseInt(parts[i]);
                } else {
                    switchIds[i] = Integer.parseInt(parts[i].substring(1));
                }
            }
            
            return new SwitchConditionalCommand(
                    new CommandGroup(""),
                    new CommandGroup(""),
                    switchIds,
                    targetValues,
                    orMode);
        }
        
        return null;
    }
    
    @Override
    public void execute() {
        for (int i = 0; i < _switchIds.length; i++) {
            boolean currentState = false;
            currentState =
                    (Interpreter.getSwitches()[_switchIds[i]] == _targetValues[i]);
            boolean stopLooping = updateState(currentState); 
            if (stopLooping) {
                break;
            }
        }
        super.execute();
    }
    
    public int[] getSwitchIds() {
        return _switchIds;
    }

}

File: src/Commands/SwitchConditionalCommand.cs
namespace Statesman.Commands
{
    public class SwitchConditionalCommand : ConditionalCommand
    {
        public static readonly string ID = "swcond";

        public int[] SwitchIds { get; }

        public SwitchConditionalCommand()
                : base()
        {
        }

        public SwitchConditionalCommand(
                CommandGroup group,
                CommandGroup elseGroup,
                int[] switchIds,
                bool[] targetValues,
                bool orMode)
                : base()
        {
            _group = group;
            _elseGroup = elseGroup;
            SwitchIds = switchIds;
            _targetValues = targetValues;
            _orMode = orMode;
        }

        public override Command CreateInstance(string[] arguments)
        {
            if (arguments.Length == 2)
            {
                string condition = arguments[1];
                bool orMode = UseOrOperator(condition);
                string[] parts = GetConditionParts(condition, orMode);
                bool[] targetValues = new bool[parts.Length];
                int[] switchIds = new int[parts.Length];

                for (int i = 0; i < parts.Length; i++)
                {
                    targetValues[i] = !parts[i].StartsWith("!");
                    if (targetValues[i])
                    {
                        switchIds[i] = int.Parse(parts[i]);
                    }
                    else
                    {
                        switchIds[i] = int.Parse(parts[i].Substring(1));
                    }
                }

                return new SwitchConditionalCommand(
                        new CommandGroup(""),
                        new CommandGroup(""),
                        switchIds,
                        targetValues,
                        orMode);
            }

            return null;
        }

        public override void Execute()
        {
            for (int i = 0; i < SwitchIds.Length; i++)
            {
                bool currentState =
                    Interpreter.Switches[SwitchIds[i]] == _targetValues[i];
                bool stopLooping = UpdateState(currentState);
                if (stopLooping)
                {
                    break;
                }
            }
            base.Execute();
        }
    }
}
